src/Networking/NexusMods.Networking.NexusWebApi/OAuth.cs

```diff
@@ -0,0 +1,270 @@
+                }
+                else
+                {
+                    AuthorizeToken(verifier, code).ContinueWith(reply =>
+                    {
+                        if (reply.Exception != null)
+                        {
+                            completionSource.SetException(reply.Exception);
+                        }
+                        else
+                        {
+                            completionSource.SetResult(reply.Result);
+                        }
+                    });
+                }
+            };
+
+        cancel.Register(() => _callbacks[state].Invoke(new OperationCanceledException(), ""));
+
+        // see https://www.rfc-editor.org/rfc/rfc7636#section-4.3
+        _os.OpenURL(GenerateAuthorizeUrl(challenge, state));
+
+        return await completionSource.Task;
+    }
+
+    /// <summary>
+    /// request a new access token
+    /// </summary>
+    /// <param name="refreshToken">the refresh token</param>
+    /// <returns>a new token reply</returns>
+    public async Task<JwtTokenReply> RefreshToken(string refreshToken)
+    {
+        var request = new Dictionary<string, string>
+        {
+            { "grant_type", "refresh_token" },
+            { "client_id", OAuthClientId },
+            { "refresh_token", refreshToken },
+        };
+
+        var content = new FormUrlEncodedContent(request);
+
+        var response = await _http.PostAsync($"{OAuthUrl}/token", content);
+        var responseString = await response.Content.ReadAsStringAsync();
+        return JsonSerializer.Deserialize<JwtTokenReply>(responseString);
+    }
+
+    private async Task<JwtTokenReply> AuthorizeToken(string verifier, string code)
+    {
+        var request = new Dictionary<string, string> {
+            { "grant_type", "authorization_code" },
+            { "client_id", OAuthClientId },
+            { "redirect_uri", OAuthRedirectURL },
+            { "code", code },
+            { "code_verifier", verifier }
+        };
+
+        var content = new FormUrlEncodedContent(request);
+
+        var response = await _http.PostAsync($"{OAuthUrl}/token", content);
+        var responseString = await response.Content.ReadAsStringAsync();
+        return JsonSerializer.Deserialize<JwtTokenReply>(responseString);
+    }
+
+    private void OnNXMUrl(NXMUrlMessage url)
+    {
+        if ((url.Value.UrlType == NXMUrlType.OAuth) && (url.Value.OAuth.State != null))
+        {
+            if (_callbacks.ContainsKey(url.Value.OAuth.State))
+            {
+                _callbacks[url.Value.OAuth.State].Invoke(null, url.Value.OAuth.Code ?? "");
+                _callbacks.Remove(url.Value.OAuth.State);
+            }
+            else
+            {
+                _logger.LogWarning("received unexpected oauth callback");
+            }
+        }
+    }
+
+    private string MakeUUID()
+    {
+        return Guid.NewGuid().ToString();
+    }
+
+    private string ToBase64(byte[] input)
```

Comment:
These would be good candidates for [extension methods](https://learn.microsoft.com/en-us/dotnet/csharp/programming-guide/classes-and-structs/extension-methods). We do Base64 conversion enough it would be nice to have them in a more central project and usable by components.